pkgs/applications/virtualization/xen/generic/default.nix

```diff
@@ -220,6 +235,56 @@ let
       )
     );
 
+  # The stubdomain also needs external tar.xz files.
+  extFiles = (
+    let
+      extFile =
+        # The extFile function takes in a `pname` attribute for the external file's name, and an extension attribute for the type of archive extension used by said file.
+        pname: extension:
+        let
+          name = "${pname}-${pkg.extFiles.${pname}.version}${extension}"; # Example output: zlib-1.2.3.tar.gz
+        in
+        {
+          src = fetchurl {
+            url = "https://xenbits.xenproject.org/xen-extfiles/${name}";
+            hash = pkg.extFiles.${pname}.hash;
+          };
+          path = "stubdom/${name}";
+        };
+    in
+    {
+      gmp = extFile "gmp" ".tar.bz2";
+      lwip = extFile "lwip" ".tar.gz";
+      newlib = extFile "newlib" ".tar.gz";
+      pciutils = extFile "pciutils" ".tar.bz2";
+      polarssl = extFile "polarssl" "-gpl.tgz";
+      tpm_emulator = extFile "tpm_emulator" ".tar.gz";
+      zlib = extFile "zlib" ".tar.gz";
```

Comment:
Do these absolutely need to be vendored? Obviously I don’t really know what’s going on here but I would much prefer to use the Nixpkgs versions of these if possible, especially if we’re having to patch their build systems again anyway. I guess we’re invoking their own Buildroot type thing to build their own Linux tree here? It’d be really nice if we could at least use our own `.src`s. Happy to hear an explanation of what’s going on here and why this is a dumb request though :)

Reply:
> Do these absolutely need to be vendored?

They do, unfortunately. MiniOS stubdomains require ancient versions of all of these libraries, and Xen patches them up as needed during the build process. I tried using nixpkgs' `.src`s[^1], but the build naturally failed as the patches Xen applies did not match the new versions' indexes.

> I guess we’re invoking their own Buildroot type thing to build their own Linux tree here?

More accurately, we're sneaking a whole new (non-Linux) kernel build in here.

> Happy to hear an explanation of what’s going on here and why this is a dumb request though

One of Xen's exclusive features is the stubdomain system. They're tiny, lightweight VMs that take away some of dom0's privilege over the system's hardware. You could have a stubdomain that runs the QEMU device model (`xen.qemu-system-i386`), removing the need for running it in dom0, thus improving the system's security by isolating this attack vector from the host.
MiniOS is one of the operating system/microkernel that runs inside a stubdomain, and is what Xen builds by default. One could also use a Linux-based or [MirageOS-based](https://github.com/NixOS/nixpkgs/pull/23955) stubdomain.

I'll try asking on XenDevel if building MiniOS with ancient dependencies is still relevant, and see what they answer.

[^1]: I tried doing this by adding a hacky function that takes the `${source}.src`, calls `tar` to package them using the correct file extensions, and copy them to the right places in the build's working directory.

Reply:
So, looks like MiniOS is considered legacy, and is being replaced by `unikraft`. We can still build it, but it should at the very least be a disabled-by-default component.

@CertainLach are you sure this is necessary for Qubes, considering ITL uses Linux-based stubdomains?

Reply:
Thanks for the detailed explanation! Having to bundle these old library versions for their bespoke system build system sucks but it sounds like it’s unavoidable at least for now if we want to package MiniOS. However, if it’s deprecated and being replaced, then I might prefer to dodge this ugliness by skipping packaging it altogether in favour of the newer stuff, as long as nothing like Qubes actually needs it.

Reply:
I'm not sure Qubes uses it. https://github.com/QubesOS/qubes-vmm-xen-stubdom-linux

If there are no objections from lach, I'd be in favour of closing this. On my personal Xen deployment, I use Linux-based stubdoms anyway.

@digitalrayne do you think there's a need for MiniOS-based stubdoms?

Reply:
> @CertainLach are you sure this is necessary for Qubes, considering ITL uses Linux-based stubdomains?

No, it is not necessary for qubes, I implemented it because I thought it is needed for qubes while it is not, and because there was a comment for stubdom make target.

Reply:
Still, I believe NixOS Xen needs to be secure OOTB, and stubdom isolation is a good thing, if MiniOS-based stubdoms are only official way to do that - I think it would be better to have it?

Reply:
I don't know about MiniOS alternatives, correct me if I'm wrong.

Reply:
> Still, I believe NixOS Xen needs to be secure OOTB, and stubdom isolation is a good thing, if MiniOS-based stubdoms are only official way to do that - I think it would be better to have it?

Linux-based stubdomains are also a valid way to ensure device isolation. Since they're a legacy component, we can assume that Xen itself doesn't want them being used anymore.

Reply:
This uses a TLS library from 2012, an IP stack from 2008, and a Zlib from 2005. All of those projects have had CVEs since those releases. I’d be more inclined to mark MiniOS with `knownVulnerabilities` than to enable it by default for security reasons.

Reply:
I agree that it is ancient, the only remaining question from me is

> Does Xen even uses it by default when built with stubdom target?

If not and it needs further configuration anyway - then sure, kill it with fire.

Reply:
> I don't know about MiniOS alternatives, correct me if I'm wrong. Does Xen even uses it by default when built with stubdom target?

All the `stubdom` target builds are the four `.gz` files listed in `passthru`. You can optionally use them to run pre-built stubdomains by calling them in an `xl` configuration.

Reply:
@SigmaSquadron - I agree that it makes sense to remove this. I think it is added maintenance burden for a fairly niche use case, I'm not aware of these being used by any default functionality, especially in the "NixOS as Dom0" use case. I also think that @emilazy makes an excellent point around the additional security overhead of carrying this and making it available. If it were ever re-added I absolutely agree it should be marked  with `knownVulnerabilities`.